### PR TITLE
Feature: tf.contrib.data.unordered_merge() #17471

### PR DESCRIPTION
**Pull request by toshirin**
Here is the feature for issue #17471.

### CODE REVIEW

**Review by deasuke — APPROVED** [review]
Is approving this is the way to add my confirmation?

**Comment by toshirin** [comment]
@mrry How is my submitted code?

**Comment by tfboyd** [comment]
@deasuke  That is good and your intent is clear.  I removed the label.  This is a good and substantial add as far as I can tell let's try to resolve these conflicts and get this in.

File: tensorflow/contrib/data/kernels/unordered_merge_dataset_op.cc [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+  class Dataset : public GraphDatasetBase {
+   public:
+    explicit Dataset(OpKernelContext* ctx,
+                     const std::vector<DatasetBase*>& inputs)
+        : GraphDatasetBase(ctx), inputs_(inputs) {
+      const auto& input = inputs_[0];
+      for (DataType dt : input->output_dtypes()) {
+        output_dtypes_.push_back(dt);
+      }
+      output_shapes_.insert(output_shapes_.end(),
```
Comment (mrry) [inline tensorflow/contrib/data/kernels/unordered_merge_dataset_op.cc]: The shape for each component here should be the "most specific compatible shape" of all the inputs. See the logic in `concatenate_dataset_op.cc` for an example:

https://github.com/tensorflow/tensorflow/blob/4984a60e7147edef532ca1b15050471e81e45841/tensorflow/core/kernels/data/concatenate_dataset_op.cc#L167

File: tensorflow/contrib/data/kernels/unordered_merge_dataset_op.cc [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        mutex_lock l(mu_);
+        if (input_impls_.empty()) {
+          *end_of_sequence = true;
+          return Status::OK();
+        }
+        out_tensors->clear();
+        out_tensors->reserve(1);
```
Comment (mrry) [inline tensorflow/contrib/data/kernels/unordered_merge_dataset_op.cc]: If reserving here, it should be `output_dtypes().size()` elements, not `1`.

File: tensorflow/contrib/data/python/ops/unordered_merge.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        output_types=nest.flatten(self._datasets[0].output_types))
+    # pylint: enable=protected-access
+
+  @property
+  def output_classes(self):
+    return self._datasets[0].output_classes
+
+  @property
+  def output_shapes(self):
+    return self._datasets[0].output_shapes
```
Comment (mrry) [inline tensorflow/contrib/data/python/ops/unordered_merge.py]: As in the C++ code, this should be the most specific compatible shape across all input datasets.

File: tensorflow/contrib/data/python/ops/unordered_merge.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        message = ("The argument to `unordered_merge()` "
+                   "must be an iterable of datasets.")
+        raise TypeError(message)
+
+    shapes0 = _shapes_to_list(self._datasets[0].output_shapes)
+    types0 = self._datasets[0].output_types
+
+    for ds in self._datasets:
+      current_shapes = _shapes_to_list(ds.output_shapes)
+      if shapes0 != current_shapes:
```
Comment (mrry) [inline tensorflow/contrib/data/python/ops/unordered_merge.py]: The shapes only need to be compatible, not necessarily equal.

File: tensorflow/contrib/data/python/ops/unordered_merge.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+def _shapes_to_list(shapes):
+  if TensorShape == type(shapes):
+    return shapes.as_list()
+  flat_shapes = nest.flatten(shapes)
+  list_shapes = [_shapes_to_list(s) for s in flat_shapes]
+  return nest.pack_sequence_as(shapes, list_shapes)
+
+class UnorderedMergeDataset(dataset_ops.Dataset):
+  """A merged `Dataset` of its input datasets without garantee of data order."""
```
Comment (mrry) [inline tensorflow/contrib/data/python/ops/unordered_merge.py]: Typo: "guarantee"

File: tensorflow/contrib/data/kernels/unordered_merge_dataset_op.cc [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        int orig_input_start = input_start;
+        do {
+          const auto& input_impl = input_impls_[input_start];
+          std::vector<Tensor> input_tensors;
+          TF_RETURN_IF_ERROR(
+              input_impl->GetNext(ctx, &input_tensors, end_of_sequence));
```
Comment (mrry) [inline tensorflow/contrib/data/kernels/unordered_merge_dataset_op.cc]: The implementation here is deterministic, which means that it's equivalent to `tf.contrib.data.sample_from_datasets(tf.data.Dataset.range(len(datasets)).repeat(None).map(lambda i: tf.one_hot(i, len(datasets))), datasets)`.

I think the original spirit of the feature request was to implement a non-deterministic merge between multiple datasets, similar to `tf.contrib.data.parallel_interleave(..., sloppy=True)`, such that elements would be produced from any available input as soon as they were ready. Can you try implementing that behavior instead?

**Review by mrry — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/data/kernels/unordered_merge_dataset_op.cc (hunk 6 above)
Comment (deasuke) [reply tensorflow/contrib/data/kernels/unordered_merge_dataset_op.cc]: The API tf.contrib.data.sample_from_datasets looks like a new one and maybe satisfy my original request -- my request **allows** (not requires) non-deterministic behavior -- if it's fast enough. I think I have to try evaluating it.

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tensorflowbutler** [comment]
It has been 44 days that this pull-request has stalled. Please create a new pull-request with the requested changes.

**Pull request closed without merging**